### PR TITLE
CPU mem reducing with cpu communication and meta_init

### PR DESCRIPTION
**Pull request by ht-zhou**

### CODE REVIEW

File: bloom-inference-scripts/test.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        # meta init
+        if rank == 0:
+                with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA], 
+                profile_memory=True, 
+                on_trace_ready=torch.profiler.tensorboard_trace_handler('./log/test'),
+                record_shapes=True,
+                with_stack=True
+                ) as prof:        
+                        model = BloomForCausalLM.from_pretrained("/data2/users/lccsr/bloom3b/data", low_cpu_mem_usage=True).half()
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:24]: 把这个路径设置成args参数吧

File: bloom-inference-scripts/test.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                        model_list = []
+                        with init_empty_weights():
+                                configuration = BloomConfig.from_json_file('/data2/users/lccsr/bloom3b/data/config.json')
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:27]: 同上，弄成统一的传入参数，要不然一个一个改起来太麻烦。

File: bloom-inference-scripts/test.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                                for i in range(world_size):
+                                        model_list.append(BloomForCausalLM(configuration).half())
+                        model_list = get_8bit_tp_model_list(model, model_list, world_size)
+                        for name, param in model_list[0].named_parameters():
+                                param_list = [param.data]
+                                for i in range(1, world_size):
+                                        param_list.append(model_list[i].state_dict()[name])
+                                param_tensor = torch.zeros_like(param_list[0],dtype=param_list[0].dtype)
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:35]: zeros_like dtype自动设置成param_list[0]的dtype了吧

File: bloom-inference-scripts/test.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        tokenizer = BloomTokenizerFast.from_pretrained("/data2/users/lccsr/bloom3b/data")
+        inputs = tokenizer("Hello, my dog is cute", return_tensors="pt").to(rank)
+        generate_kwargs = dict(max_new_tokens=100, do_sample=False)
+        outputs = model.generate(**inputs, **generate_kwargs)
+        outputs = tokenizer.batch_decode(outputs, skip_special_tokens=True)
+        print(rank, outputs)
+                
+
+    
+def run16():
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:81]: run_fp16

File: bloom-inference-scripts/test.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import torch
+from transformers import BloomTokenizerFast, BloomForCausalLM, BloomConfig
+import torch.distributed as dist
+import os
+from utils import replace_8bit_linear_tp, get_8bit_tp_model, get_8bit_tp_model_list, replace_8bit_linear, getModelSize, init_empty_weights, Linear8bitTP
+import time
+import torch.profiler
+       
+def run8():
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:9]: run_int8

File: bloom-inference-scripts/test.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        world_size = torch.cuda.device_count()
+        local_rank = int(os.getenv("LOCAL_RANK", "0"))
+        rank = local_rank
+        dist.init_process_group(backend='nccl', world_size=world_size, rank=rank)
+        cpu_group = dist.new_group(backend='gloo') if dist.get_backend() != 'gloo' else None
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:14]: cpu_group = dist.new_group(backend='gloo')
if else在此没有任何意义

File: bloom-inference-scripts/test.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
+import torch
+from transformers import BloomTokenizerFast, BloomForCausalLM, BloomConfig
+import torch.distributed as dist
+import os
+from utils import replace_8bit_linear_tp, get_8bit_tp_model, get_8bit_tp_model_list, replace_8bit_linear, getModelSize, init_empty_weights, Linear8bitTP
+import time
+import torch.profiler
+       
+def run8():
+        world_size = torch.cuda.device_count()
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:10]: 一般都是 world_size = torch.distributed.get_world_size()
因为可以再4GPU上启动2个process，你这么写会报错

File: bloom-inference-scripts/test.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+                on_trace_ready=torch.profiler.tensorboard_trace_handler('./log/test'),
+                record_shapes=True,
+                with_stack=True
+                ) as prof:
+                        with init_empty_weights():
+                                configuration = BloomConfig.from_json_file('/data2/users/lccsr/bloom3b/data/config.json')
+                                model = BloomForCausalLM(configuration).half()
+                        model = get_8bit_tp_model(model, rank, world_size)
+                        for name, param in model.named_parameters():
+                                param_tensor = torch.zeros(param.data.size(), dtype=param.dtype)
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:55]: 用torch.empty_like?

File: bloom-inference-scripts/test.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                                dist.scatter(param_tensor, src=0, group=cpu_group)
+                                param = torch.nn.Parameter(param_tensor, requires_grad=False)
+                                name_list = name.split('.')
+                                module = model._modules[name_list[0]]
+                                for i in range(1, len(name_list) - 1):
+                                        module = module._modules[name_list[i]]
+                                module._parameters[name_list[-1]] = param
```
Comment (feifeibear) [inline bloom-inference-scripts/test.py:62]: 这几行代码可以抽取成一个函数。

Comment (feifeibear) [reply bloom-inference-scripts/test.py:62]: 以后可以复用

File: bloom-inference-scripts/utils.py [hunk 10]
```diff
@@ -5,7 +5,24 @@
...
+    for param in model.parameters():
+        param_size += param.nelement() * param.element_size()
+        param_sum += param.nelement()
+    buffer_size = 0
+    buffer_sum = 0
+    for buffer in model.buffers():
+        buffer_size += buffer.nelement() * buffer.element_size()
+        buffer_sum += buffer.nelement()
+    all_size = (param_size + buffer_size) / 1024 / 1024
+    print('模型总大小为：{:.3f}MB'.format(all_size))
```
Comment (feifeibear) [inline bloom-inference-scripts/utils.py:24]: 用英文吧。

File: bloom-inference-scripts/utils.py [hunk 11]
```diff
@@ -232,76 +254,118 @@ def replace_8bit_linear(model, threshold=6.0, modules_to_not_convert="lm_head"):
     return model
 
 def get_8bit_tp_model(model, rank, world_size):
+    model = replace_8bit_linear_tp(model)
     for name, module in model.named_modules():
         if isinstance(module, Linear8bitTP):
+            bias_list = list(module.bias.data.chunk(world_size, dim=0))
```
Comment (feifeibear) [inline bloom-inference-scripts/utils.py:260]: 思睿觉得这块需要clone().detach()一下。因为chunk是可以autograd的，所以chunk前的data还会保留在计算图里。

Comment (feifeibear) [reply bloom-inference-scripts/utils.py:260]: 可能在def get_8bit_tp_model上面加一个decorator可以有类似效果
@torch.no_grad()

File: bloom-inference-scripts/utils.py [hunk 12]
```diff
@@ -232,76 +254,118 @@ def replace_8bit_linear(model, threshold=6.0, modules_to_not_convert="lm_head"):
...
+            
             delattr(module, "bias")
             setattr(module, "bias", nn.Parameter(bias))
             
+            
         if isinstance(module, EmbeddingTP):   
             weight_list = list(module.weight.chunk(world_size, dim=1))
-            weight = nn.Parameter(weight_list[rank].clone().detach())
             delattr(module, 'weight')
+            weight = nn.Parameter(weight_list[rank])
```
Comment (feifeibear) [inline bloom-inference-scripts/utils.py:285]: weight_list[rank].clone().detach()?

File: bloom-inference-scripts/test.py (hunk 7 above)
Comment (ht-zhou) [reply bloom-inference-scripts/test.py:10]: 这个地方不对吧，get_world_size需要有process group才行，或者就是先设定一个world_size的数值，然后再init_process_group

File: bloom-inference-scripts/test.py (hunk 8 above)
Comment (ht-zhou) [reply bloom-inference-scripts/test.py:55]: 不行，因为好像会得到一个meta tensor会出问题，我看了一下这里增加的mem可以忽略

File: bloom-inference-scripts/test.py (hunk 7 above)
Comment (feifeibear) [reply bloom-inference-scripts/test.py:22]: 这个可以启动时候控制
torchrun --nproc_per_node=X

File: bloom-inference-scripts/test.py (hunk 8 above)
Comment (feifeibear) [reply bloom-inference-scripts/test.py:70]: OK

**Review by feifeibear — APPROVED** [review]
